### PR TITLE
added management command

### PR DESCRIPTION
**Pull request by jrief**
Can this command be useful?

It sometimes happens, that I have orphaned files in my folder `filer_public`, which do not have any associated meta data. With `./manage.py orphaned_files` we can list them, and with `./manage.py orphaned_files --delete` we can delete them.

### CODE REVIEW

**Comment by yakky** [comment]
@jrief :+1:

**Comment by yakky** [comment]
@jrief I'd just an option to only output filenames instead of "Found orphaned file...", to be able to pipe the output somewhere (I'd use verbosity to control the output form)

**Comment by jrief** [comment]
Then the command should be renamed to, say ``filer_check`` and can be used to not only check for orphaned files, but also for missing files. How should we name the options? My proposal is ``--orphans-only``, ``--missing-only`` and ``--delete-orphans``.

**Comment by yakky** [comment]
I'd use multiple options:
``--orphans``: scan for orphans
``-missing``: scan for missing
``--delete-orphans``: delete orphans
``--verbose=3``: output verbose string instead of raw filenames

the first two can be combined together to scan both

**Comment by yakky** [comment]
@jrief would you have a look at my comment above?

**Comment by wiesson** [comment]
Nice idea but it did not work for me because I changed the filer urls in the django settings. WIth `prefix = settings.FILER_STORAGES['public']['main']['UPLOAD_TO_PREFIX']` it worked perfectly.

**Comment by jrief** [comment]
@wiesson thanks for the hint, but your approach only works if FILER_STORAGES has been configured explicitly in the project settings. I therefore will wrap it into a try/except-block to fallback to the defaults.

**Comment by yakky** [comment]
@jrief any progress on this?

File: filer/management/commands/orphaned_files.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    def handle(self, *args, **options):
+        from filer.models.filemodels import File
+
+        def walk(absdir, reldir):
+            child_dirs, files = self.storage.listdir(absdir)
+            for filename in files:
+                relfilename = os.path.join(reldir, filename)
+                try:
+                    File.objects.get(file=relfilename)
```
Comment (czpython) [inline filer/management/commands/orphaned_files.py]: why not `if not File.objects.filter(file=relfilename).exists()`

**Comment by jrief** [comment]
@yakky thanks for reminding me and sorry for not proceeding on this. I'm stuck with far too many other issues.

**Comment by benzkji** [comment]
just curious, does this work if for example S3 storage is configured? Also, does it respect private/public flag on files? Methinks not, as it just uses DefaultStorage() ?

**Comment by jrief** [comment]
@yakky 
implemented as you proposed, command is named ``filer_check`` with these options:
* ``--orphans``: scan for orphans
* ``--delete-orphans``: delete orphans
* ``--missing``: scan for missing
* ``--delete-missing``: delete missing
* ``--no-input``/``--noinput`` to skip safety question

@czpython 
thanks for the hint

added unit tests for all command options and entry in CHANGELOG.

sorry for the long delay

File: filer/management/commands/orphaned_files.py (hunk 1 above)
Comment (jrief) [reply filer/management/commands/orphaned_files.py]: fixed

**Comment by jrief** [comment]
Could someone please explain me, what's wrong with this [failing test](https://travis-ci.org/divio/django-filer/jobs/292451336)?

My local ``isort`` does not complain about my import sorting, and from Travis's output I don't get any feedback.

**Comment by jrief** [comment]
@czpython fixed you annotation

**Comment by jrief** [comment]
@czpython all annotations have been fixed

**Comment by jrief** [comment]
@yakky conflicts have been resolved, but Travis-CI does not want to finish :(

**Comment by yakky** [comment]
@jrief travis has completed, but github still says there are conflicts

**Comment by jrief** [comment]
@yakky GitHub tells me in big bold letters: **This branch has no conflicts with the base branch**.

Do you really get a different message? Then that must be a GitHub bug.

**Comment by yakky** [comment]
@jrief apparently it is
This is what I see :smile: 
image

Applying it manually I have a conflict on the changelog file
Can you try to fetch latest develop (commit id eccd8533de2b76a6793d294b917854eb27dbccae ) and try to merge it locally

File: docs/management_commands.rst [hunk 2]
```diff
@@ -12,3 +12,33 @@ command.
...
+
+Invoking::
+
+    ./manage.py filer_check --delete-missing
+
+deletes those file references from the database.
+
+Invoking::
+
+    ./manage.py filer_check --orphanes
```
Comment (czpython) [inline docs/management_commands.rst]: typo

File: docs/management_commands.rst [hunk 3]
```diff
@@ -12,3 +12,33 @@ command.
...
+
+lists all files found on disk belonging to the configured storage engine, but which
+are not referenced by the database.
+
+Invoking::
+
+    ./manage.py filer_check --delete-orphanes
```
Comment (czpython) [inline docs/management_commands.rst]: typo

File: filer/management/commands/filer_check.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+from django.core.files.storage import DefaultStorage
+from django.core.management.base import BaseCommand, CommandError
+from django.utils.module_loading import import_string
+from django.utils.six.moves import input
+from django.utils.translation import ugettext_lazy as _
+from filer.settings import DEFAULT_FILER_STORAGES
+
+
+class Command(BaseCommand):
```
Comment (czpython) [inline filer/management/commands/filer_check.py]: management commands shouldn't use mark strings for translation.

File: filer/management/commands/filer_check.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    def handle(self, *args, **options):
+        if options['missing']:
+            self.verify_references(options)
+        if options['delete_missing']:
+            if options['interactive']:
+                msg = _("\nThis will delete entries from your database. Are you sure you want to do this?\n\n"
+                        "Type 'yes' to continue, or 'no' to cancel: ")
+                if input(msg) != 'yes':
+                    raise CommandError("Aborted: Delete missing file entries from database.")
```
Comment (czpython) [inline filer/management/commands/filer_check.py]: why raise an exception, no is a valid option

File: filer/management/commands/filer_check.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            self.verify_references(options)
+
+        if options['orphans']:
+            self.verify_storages(options)
+        if options['delete_orphans']:
+            if options['interactive']:
+                msg = _("\nThis will delete orphaned files from your storage. Are you sure you want to do this?\n\n"
+                        "Type 'yes' to continue, or 'no' to cancel: ")
+                if input(msg) != 'yes':
+                    raise CommandError("Aborted: Delete orphaned files from storage.")
```
Comment (czpython) [inline filer/management/commands/filer_check.py]: same as above.

File: docs/management_commands.rst (hunk 2 above)
Comment (jrief) [reply docs/management_commands.rst]: fixed

File: docs/management_commands.rst (hunk 3 above)
Comment (jrief) [reply docs/management_commands.rst]: fixed

File: filer/management/commands/filer_check.py (hunk 4 above)
Comment (jrief) [reply filer/management/commands/filer_check.py]: fixed

File: filer/management/commands/filer_check.py (hunk 5 above)
Comment (jrief) [reply filer/management/commands/filer_check.py]: fixed

File: filer/management/commands/filer_check.py (hunk 6 above)
Comment (jrief) [reply filer/management/commands/filer_check.py]: fixed

**Comment by jrief** [comment]
@yakky merged HEAD of branch develop (eccd8533de2b76a6793d294b917854eb27dbccae) into this branch, with no side effect.

**Comment by jrief** [comment]
Unit test job 2196.35 is failing for throttling reasons:

> Access denied | packages.npmjs.org used Cloudflare to restrict access

Please restart the job.

**Comment by jrief** [comment]
Ping @czpython 
All request changes have been fixed. Is there still anything missing?

**Comment by peterfarrell** [comment]
@yakky is there anything I can do to help get this merged and released?

**Comment by jrief** [comment]
Today I had to check the consistency of files in one of my projects. Unfortunately I was quite disappointed that django-filer still lacks this feature, although this pull request is pending and waiting to be merged for one year now.

**Comment by jrief** [comment]
Now I rebased this pull request upon the current `development` branch, so it should be ready for the next version.

I really would appreciate, if in version 1.6 this useful feature finally would be available, almost 3 years after starting to work on it.

**Comment by peterfarrell** [comment]
@jrief I am at the point where I'm debating forking the project with this feature and releasing it on PyPi. I know I'm not the only person with the need for this functionality.  I do not want to be forced to take that option because it's fragmentation, however the time is approaching quickly.

**Comment by jrief** [comment]
Today this pull requests celebrates its 🎂**3rd birthday** 🎂and is still unmerged!

Dear maintainers of django-filer: If you need help with maintenance, please consider to add me to the list of core maintainer of this project.

**Comment by FinalAngel** [comment]
@jrief thanks for the PR and changes requested from various reviewers. It gets merged today :)

I also added you to the collaborators, I appreciate the help on Filer 🙏

**Pull request merged**